cylc8 first-pass

oliver-sanders opened the pull request:
Sibling PRs: https://github.com/cylc/cylc-sphinx-extensions/pull/29, https://github.com/cylc/cylc-flow/pull/3777

**A quick first-pass of the documentation:**

* Reference configurations properly (`:cylc:conf:`).
* Fix some Cylc7/8 changes.
* Remove some obsolete documentation.
* Make code examples fix the suite design guide.
* Auto-document more stuff from Cylc Flow.
* Simplify some wording.
* Correct some syntax highlighting.
* Consistently capitilise Cylc (except the `cylc` command).
* Merge "architecture" in with the "introduction" to allow the intro section to grow.

This started out as a quick pass to use the cylc-domain e.g. `[runtime]` -> `:cylc:conf:'[runtime]'`, however, it is very hard to edit just one thing in the documentation at the moment :/

**The Grand Scheme Of Things:**

This isn't the "great edit" that the docs need, however, it should make that edit a bit easier when it comes.

These docs aren't polished ready for release, however, they are a *bit* better than before.

Most of the non-indentation changes are in the user guide, namely writing and running suites.

**Pre 8.0.0:**

Here's a non-exhaustive list of stuff to do to the docs pre-cylc8:

* Terminology changes (e.g. suite->flow).
* Platformise stuff after the `[platforms]` changes.
* Review the configuration documentation in cylc-flow.
* Review all the docs ;)
* Fix the tutorial to Cylc8 changes.
* Write a Cylc7->8 transition guide.
* Fix the screenshots as well as all graph / GUI references.
* Amend installation instructions as they change + write quickstart.
* Break apart some of the monolithic sections (e.g. writing / running suites) into more digestible parts.

oliver-sanders commented:
Tests will fail until the next version of cylc-sphinx-extensions is released.

MetRonnie commented on src/glossary.rst:
```diff
@@ -1,18 +1,23 @@
    suite directory
       The suite directory contains all of the configuration for a suite e.g.
-      the ``suite.rc`` file.
+      the :cylc:conf:`suite.rc` file.
```
Stuff like this is going to conflict every single time with my #142 PR 😅

oliver-sanders replied:
We'll try to get yours in first then!

oliver-sanders replied:
For now yes. Ronnies change should be merged soon so I'll get the rebase!

MetRonnie commented on src/user-guide/writing-suites.rst:
```diff
@@ -3,94 +3,62 @@
+:cylc:conf:`suite.rc`
+   The file which configures the workflow.
+``bin/`` (optional)
+   A directiory where you can put scripts and executables for use
```
```suggestion
   A directory where you can put scripts and executables for use
```

hjoliver commented on src/user-guide/suite-storage-etc.rst:
```diff
@@ -3,14 +3,14 @@
-preferred revision control systems and suite meta-data requirements that
-are difficult to anticipate. We can, however, recommend the use of
-*Rose* to do all of this very easily and elegantly with cylc suites.
+Cylc suites should be portable and may be version controlled, however, beyond
+this Cylc does not have a built-in solution for suite storage and discovery,
```
```suggestion
Cylc does not have a built-in solution for suite storage and discovery,
``` 
("should be portable"?)

oliver-sanders replied:
I didn't feel that this was adding anything but can add it back.

hjoliver commented on src/user-guide/writing-suites.rst:
```diff
@@ -1300,17 +1288,17 @@ with open arrow heads:
 """"""""""""""""
 
+.. tutorial:: Suicide Trigger Tutorial <tut-cylc-suicide-triggers>
+
```
```suggestion
.. TODO: gut this section in light of SoD changes (suicide triggers hardly needed now)
```

hjoliver commented on src/user-guide/writing-suites.rst:
```diff
@@ -2915,8 +2899,8 @@ Parameterized Cycling
-:ref:`Workflows For Cycling Systems`.  For most purposes use of
-a proper *cycling workflow* is recommended, wherein cylc incrementally
+:ref:`Workflows For Cycling Systems`. For most purposes use of
+a proper :term:`cycling` workflow is recommended, wherein Cylc incrementally
```
Maybe insert a TODO reminder in here.
```
.. TODO: update this section in light of SoD internals: parameterized cycling is nowhere near as 
inefficient as it was, at run time, although it still generates many more task definitions that have
 to be held in memory.
```

oliver-sanders commented on cylc/doc/etc/tutorial/retries-tutorial/suite.rc:
```diff
@@ -10,14 +10,14 @@
+            else
+               exit 1
+            fi
+        """
```
> Hmmm, I don't recall that we agreed on this indentation thing.

😁

Besides readibility issues there was a lot of inconsistency with indentation and code examples weren't always consistent with the suite design guide.

> even though I'm obviously right about it!!!

🤮

> But we should at least put a comment in the user guide to explain why it is OK to indent the lines of a bash script

True, there is a note in the suite design guide to that effect but we should probably add a note to the tutorial and writing suites sections if not already present.

> and when it occasionally might not work.

Since the dedent change indenting scripts works as expected even with indentation-sensitive stuff like heredocs or Python scripts:

```console
$ cat x/suite.rc
[scheduling]
    [[graph]]
        R1 = foo
[runtime]
    [[foo]]
        script = """
            cat >> log.txt <<_EOF
            a
            b
            _EOF
        """
$ cat x/work/1/foo/log.txt 
a
b
```

Are there any remaining cases where it can fail?

hjoliver replied:
> Are there any remaining cases where it can fail? 

Meh, I can't think of any.  Here docs are the obvious problem for bash job scripts, but dedent solved that.  

https://github.com/cylc/cylc-doc/issues/149

oliver-sanders commented on src/introduction/how-cylc-works.rst:
```diff
@@ -342,14 +342,6 @@ their own dependencies so that optimal scheduling, as described in the
-
 .. [1] Future plans for EcoConnect include additional deterministic regional
        weather forecasts and a statistical ensemble.
 .. [2] Note that simply overlapping the single cycle point schedules of the
```
Fair enough, it's full of NWP context too, happy to remove the content and leave the title or else leave the whole thing pending re-write.

hjoliver replied:
Let's remove it.

oliver-sanders commented on src/suite-design-guide/general-principles.rst:
```diff
@@ -106,7 +106,7 @@ Installing Files At Start-up
 Use ``rose suite-run`` *file creation mode* or ``R1``
 install tasks to copy files to the self-contained suite run directory at
-start-up.  Install tasks are preferred for time-consuming installations because
+start-up. Install tasks are preferred for time-consuming installations because
```
Yep, this was common, personally I find it really annoying as better solution is a typographical one. However, nowerdays most markup languages including HTML, markdown and RST will ignore surplus whitespace so these spaces are no longer functional (even if we built PDF via LaTeX the whitespace would still get stripped).

hjoliver replied:
> Yep, this was common, personally I find it really annoying as the better solution is a typographical one, it seems mad to expect people to edit their text to overcome typesetting deficiencies.

It goes back to physical typewriters, typing on paper!

oliver-sanders commented on src/user-guide/remote-job-management.rst:
```diff
@@ -3,350 +3,63 @@ Remote Job Management
   written by the job submit command, and read by job poll and kill commands
-  (all over SSH)
-
-A Concrete Example
```
Yep, stripping for now as it's out of date with recent changes. We have a "security" document, we could move it there and possibly even auto-generate it from flow log files.

oliver-sanders commented on src/user-guide/writing-suites.rst:
```diff
@@ -1063,7 +1044,7 @@ Multiple integer exclusions are also valid in the same way as the syntax
+.. code-block:: sub
 
    R/P1!(2,3,7)  # Run with step 1 to the final cycle point,
                  # but not at points 2, 3, or 7.
```
FYI the language "sub" is something I created for Rose for code examples which aren't actually bash but where you want comments to be highlighted. It also highlights `<substitutions>` (hence the name).

oliver-sanders commented on src/user-guide/running-suites.rst:
```diff
@@ -285,22 +288,20 @@ for task messaging by TCP or SSH. See :ref:`Polling To Track Job Status`.
 
 - task-to-suite messaging via TCP (using ZMQ protocol)
-- task-to-suite messaging via non-interactive SSH to the suite host,
-  then local TCP
```
Yes, however, we haven't restored it yet (I think still waiting on a decision over whether we can rely on TCP forwarding or whether we should re-implement the Cylc7 SSH-forwarding logic).

hjoliver replied:
Pretty sure we decided we had to re-implement the old logic - ping @dpmatthews ?   Maybe bang a TODO in here at least, so we don't forget to come back and update this section.

dpmatthews replied:
Yes, I think we've decided to use the old method. See cylc/cylc-flow#3327.

oliver-sanders commented:
Obviously, don't merge this branch until the tests pass.

oliver-sanders commented:
Added an extra couple of commits onto the end.

I'm not quite sure what happened here but it looks like the migration of extensions into cylc-sphinx-extensions was never properly completed. I must have been in a rush. There were some static files left behind in `src/_static`.

This was covering up a much bigger issue to do with the installation of static files, which, sadly was not working with the intended setup.

So I've added a commit onto the end of https://github.com/cylc/cylc-sphinx-extensions/pull/29 which manually copies static files across into the build directory. It would appear that this is the only real way to do it, this is what other extensions do.

MetRonnie commented:
Tried to test by checking out all the relevant branches on this, cylc-flow and cylc-sphinx-ext but the conf changes are making it a headache, Can we merge #142  first?

oliver-sanders commented:
Tests pass, took a few PRs to cylc-sphinx-extensions, cylc-doc and cylc-flow, but we are now Sphinx3, `[platforms]` and `*.cylc` compliant.

oliver-sanders commented:
Merging with two approvals.
